File: engine/compose/src/commonMain/kotlin/dev/getmilano/runtime/MilanoQuickStart.kt
package dev.getmilano

/**
 * The quick path's construction: engine, registry, and builder in one
 * call, with declared state synthesized as zero-values so a first
 * integration is a single composable. The full architecture (shared
 * engine, explicit providers) remains the recommended shape for real apps.
 */
internal fun milanoQuickBuilder(
    documentText: String,
    vocabularyJson: String,
    renderers: Map<String, MilanoRenderer>,
    context: Map<String, MilanoValue>,
    state: Map<String, MilanoValue>,
    onAction: (suspend (MilanoAction) -> MilanoValue?)?,
): MilanoViewBuilder {
    val registry = MilanoRegistry()
    for ((type, renderer) in renderers) registry.register(type, renderer)
    val engine = MilanoEngine(vocabularyJson, registry)
    val builder =
        engine
            .viewBuilder(documentText)
            .context(context)
            .stateDataProvider { declarations -> synthesizedState(declarations, state) }
    onAction?.let { handler -> builder.actionHandler { action -> handler(action) } }
    return builder
}

/**
 * Zero-values per declaration, overridden by supplied values: false, 0,
 * 0.0, empty string; null for optionals; empty arrays; records recursed.
 */
internal fun synthesizedState(
    declarations: Map<String, MilanoType>,
    supplied: Map<String, MilanoValue>,
): Map<String, MilanoValue> = declarations.mapValues { (key, type) -> supplied[key] ?: zeroValue(type) }

private fun zeroValue(type: MilanoType): MilanoValue =
    when {
        type.optional -> {
            MilanoValue.Null
        }

        type.kind is MilanoType.Kind.Bool -> {
            MilanoValue.BoolValue(false)
        }

        type.kind is MilanoType.Kind.Int -> {
            MilanoValue.IntValue(0)
        }

        type.kind is MilanoType.Kind.Double -> {
            MilanoValue.DoubleValue(0.0)
        }

        type.kind is MilanoType.Kind.Text -> {
            MilanoValue.StringValue("")
        }

        type.kind is MilanoType.Kind.Array -> {
            MilanoValue.ArrayValue(emptyList())
        }

        else -> {
            val fields = (type.kind as MilanoType.Kind.Record).fields
            MilanoValue.RecordValue(fields.mapValues { (_, field) -> zeroValue(field) })
        }
    }

File: engine/swiftui/Sources/MilanoSDK/Runtime/MilanoQuickStart.swift
import Foundation

/// The quick path's construction: engine, registry, and builder in one
/// call, with declared state synthesized as zero-values so a first
/// integration is a single view. The full architecture (shared engine,
/// explicit providers) remains the recommended shape for real apps.
enum MilanoQuickStart {

    static func builder(
        document: Data,
        vocabulary: Data,
        renderers: [String: any MilanoRenderer],
        context: [String: MilanoValue],
        state: [String: MilanoValue],
        onAction: (@Sendable (MilanoAction) async throws -> MilanoValue?)?
    ) throws -> MilanoViewBuilder {
        var registry = MilanoRegistry()
        for (type, renderer) in renderers {
            registry.register(renderer, for: type)
        }
        let engine = try MilanoEngine(vocabularyJSON: vocabulary, registry: registry)
        let builder = engine.viewBuilder(document: document)
        builder.context(context)
        builder.stateData { declarations in
            synthesized(declarations, overriding: state)
        }
        if let onAction {
            builder.actionHandler(onAction)
        }
        return builder
    }

    /// Zero-values per declaration, overridden by supplied values: false,
    /// 0, 0.0, empty string; null for optionals; empty arrays; records
    /// recursed.
    static func synthesized(
        _ declarations: [String: MilanoType], overriding supplied: [String: MilanoValue]
    ) -> [String: MilanoValue] {
        var values: [String: MilanoValue] = [:]
        for (key, type) in declarations {
            values[key] = supplied[key] ?? zeroValue(for: type)
        }
        return values
    }

    private static func zeroValue(for type: MilanoType) -> MilanoValue {
        if type.optional { return .null }
        switch type.kind {
        case .bool: return .bool(false)
        case .int: return .int(0)
        case .double: return .double(0)
        case .string: return .string("")
        // The zero-value of an enum is its alphabetically first member:
        // deterministic, and always a valid member.
        case .enumeration(let members): return .string(members.sorted()[0])
        case .array: return .array([])
        case .record(let fields): return .record(fields.mapValues(zeroValue(for:)))
        }
    }
}
